ostd/src/mm/page_table/cursor/locking.rs

```diff
@@ -0,0 +1,314 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! Implementation of the locking protocol.
+
+use core::{marker::PhantomData, ops::Range, sync::atomic::Ordering};
+
+use align_ext::AlignExt;
+
+use super::Cursor;
+use crate::{
+    mm::{
+        nr_subpage_per_huge, paddr_to_vaddr,
+        page_table::{
+            load_pte, page_size, pte_index, Child, MapTrackingStatus, PageTable,
+            PageTableEntryTrait, PageTableGuard, PageTableMode, PageTableNodeRef,
+            PagingConstsTrait, PagingLevel,
+        },
+        Paddr, Vaddr,
+    },
+    task::disable_preempt,
+};
+
+pub(super) fn lock_range<'a, M: PageTableMode, E: PageTableEntryTrait, C: PagingConstsTrait>(
+    pt: &'a PageTable<M, E, C>,
+    va: &Range<Vaddr>,
+    new_pt_is_tracked: MapTrackingStatus,
+) -> Cursor<'a, M, E, C> {
+    // Start RCU read-side critical section.
+    let preempt_guard = disable_preempt();
+
+    // The re-try loop of finding the sub-tree root.
+    //
+    // If we locked a stray node, we need to re-try. Otherwise, although
+    // there are no safety concerns, the operations of a cursor on an stray
+    // sub-tree will not see the current state and will not change the current
+    // state, breaking serializability.
+    let mut subtree_root = 'retry: loop {
+        let Some(subtree_root) = try_traverse_and_lock_subtree_root(pt, va, new_pt_is_tracked)
+        else {
+            continue 'retry;
+        };
```

Comment:
This is shorter?
```rust
loop {
    if Some(subtree_root) = .. {
        break subtree_root;
    }
}
```

In this particular case, since the loop only contains two or three lines, I prefer to just remove the "retry" tag. It's just too obvious. (Not to mention that there are even detailed comments claiming that this is a retry loop.)